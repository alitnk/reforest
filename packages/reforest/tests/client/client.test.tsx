import * as React from "react"
import { waitFor, render } from "@testing-library/react"
import "@testing-library/jest-dom"

import { useTree, useTreeData, useTreeEffect } from "../../index"

test("renders a simple list of items with the correct indexes", async () => {
  function Item({ children, value }: { children: React.ReactNode; value: string }) {
    const tree = useTree(children)
    const data = useTreeData(React.useMemo(() => ({ value }), [value]))

    return <div data-testid={data?.indexPathString}>{tree.children}</div>
  }

  function ItemList({ children }: { children: React.ReactNode }) {
    const tree = useTree(children)

    return <>{tree.children}</>
  }

  await waitFor(() => {
    const { queryByTestId } = render(
      <React.Suspense fallback={null}>
        <ItemList>
          <Item value="apple">Apple</Item>
          <Item value="orange">Orange</Item>
          <Item value="banana">Banana</Item>
        </ItemList>
      </React.Suspense>
    )

    expect(queryByTestId("1")).toHaveTextContent("Orange")
  })
})

test("renders a complex list of items with the correct indexes", async () => {
  const handleTreeUpdate = jest.fn()

  function Item({ children, value }: { children: React.ReactNode; value: string }) {
    const tree = useTree(children)
    const data = useTreeData(React.useMemo(() => ({ value }), [value]))

    return <div data-testid={data?.indexPathString}>{tree.children}</div>
  }

  function ItemList({ children }: { children: React.ReactNode }) {
    const tree = useTree(children)

    useTreeEffect(tree.treeMap, handleTreeUpdate)

    return tree.children
  }

  await waitFor(() => {
    const { queryByTestId } = render(
      <React.Suspense fallback={null}>
        <ItemList>
          <Item value="apples">
            Apples
            <Item value="fuji">Fuji</Item>
            <Item value="gala">Gala</Item>
            <Item value="honeycrisp">Honeycrisp</Item>
          </Item>
          <Item value="apples">
            Oranges
            <Item value="mandarin">Mandarin</Item>
            <Item value="naval">Naval</Item>
            <Item value="tangerine">Tangerine</Item>
          </Item>
          <Item value="pears">
            Pears
            <Item value="anjou">Anjou</Item>
            <Item value="asian">Asian</Item>
            <Item value="bosc">Bosc</Item>
          </Item>
        </ItemList>
      </React.Suspense>
    )

    expect(queryByTestId("2.3")).toHaveTextContent("Bosc")

    expect(handleTreeUpdate).toHaveBeenCalledTimes(1)

    expect(handleTreeUpdate).toMatchSnapshot()
  })
})
