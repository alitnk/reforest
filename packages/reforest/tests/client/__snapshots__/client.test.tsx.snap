// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a complex list of items with the correct indexes 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "value": "fuji",
              },
              Object {
                "value": "gala",
              },
              Object {
                "value": "honeycrisp",
              },
            ],
            "value": "apples",
          },
          Object {
            "children": Array [
              Object {
                "value": "mandarin",
              },
              Object {
                "value": "naval",
              },
              Object {
                "value": "tangerine",
              },
            ],
            "value": "apples",
          },
          Object {
            "children": Array [
              Object {
                "value": "anjou",
              },
              Object {
                "value": "asian",
              },
              Object {
                "value": "bosc",
              },
            ],
            "value": "pears",
          },
        ],
      },
      Map {
        "0.1" => Object {
          "value": "fuji",
        },
        "0.2" => Object {
          "value": "gala",
        },
        "0.3" => Object {
          "value": "honeycrisp",
        },
        "0" => Object {
          "value": "apples",
        },
        "1.1" => Object {
          "value": "mandarin",
        },
        "1.2" => Object {
          "value": "naval",
        },
        "1.3" => Object {
          "value": "tangerine",
        },
        "1" => Object {
          "value": "apples",
        },
        "2.1" => Object {
          "value": "anjou",
        },
        "2.2" => Object {
          "value": "asian",
        },
        "2.3" => Object {
          "value": "bosc",
        },
        "2" => Object {
          "value": "pears",
        },
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
